Search context: Add search context url param and append context to queries

novoselrok opened the pull request:
Adds `context` param to the URL which corresponds to the selected search context in the dropdown. If a context filter is not present in the query, then the selected search context gets appended to the query.

When switching between search contexts we automatically re-run the search (similar behaviour as the version contexts dropdown).

limitedmage commented on client/shared/src/util/url.ts:
```diff
@@ -621,6 +622,10 @@ export function buildSearchURLQuery(
         searchParameters.set('c', versionContext)
     }
 
+    if (searchContextSpec) {
+        searchParameters.set('context', searchContextSpec)
+    }
+
```
Just to confirm, In the URL scheme, is `context` going to be a separate parameter like `patternType` and not appended to the search query `q` parameter, right?

novoselrok replied:
Correct, context URL param corresponds to the context selected in the context menu dropdown.

limitedmage commented on client/web/src/SourcegraphWebApp.tsx:
```diff
@@ -322,7 +322,10 @@ class ColdSourcegraphWebApp extends React.Component<SourcegraphWebAppProps, Sour
 
         this.subscriptions.add(
             fetchSearchContexts.subscribe(contexts => {
-                this.setState({ availableSearchContexts: contexts })
+                const parsedSearchURL = parseSearchURL(window.location.search)
+                this.setState({ availableSearchContexts: contexts }, () => {
+                    this.setSelectedSearchContextSpec(parsedSearchURL.searchContextSpec || '')
+                })
             })
         )
```
It would be a good idea to write an integration tests to verify this, mocking out the call to `fetchSearchContexts`

limitedmage commented on client/web/src/search/index.tsx:
```diff
@@ -110,11 +121,16 @@ export function parseSearchURL(
         finalQuery = caseSensitive ? `${finalQuery} case:yes` : finalQuery
     }
 
+    if (appendContextFilter && searchContextSpec && !isContextFilterInQuery(finalQuery)) {
```
Super minor: I guess in this case we're pre-pending it rather than appending it? Does it make a difference to the backend?

novoselrok replied:
AFAIK, it shouldn't. As long as there aren't existing contexts in the query (which we prevent), it shouldn't matter.

limitedmage commented on client/web/src/search/stream.ts:
```diff
@@ -474,11 +476,18 @@ function search({
     trace,
 }: StreamSearchOptions): Observable<SearchEvent> {
     return new Observable<SearchEvent>(observer => {
+        console.trace()
```
```suggestion
```

limitedmage reviewed (approved):
Looks great! Left some minor suggestions.

limitedmage commented:
Just remembered that we also likely need to add code to react to changes to the URL in `Layout.tsx`: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/Layout.tsx#L163:1-164:1

novoselrok commented:
@limitedmage Lots of files changed, but mostly tests and passing the search context props everywhere. I considered splitting the PR, but there isn't a clean separation and would probably just cause more work.

quinnkeast reviewed (changes requested):
Overall looks good to me, but I think there's one bug:

If I don't have a context set at all, it's not defaulting to global or the user context, and instead shows an empty context selector:

![image](https://user-images.githubusercontent.com/6304497/107951871-48621900-6f99-11eb-9f7c-20781dccc884.png)

limitedmage commented on client/web/src/search/input/SearchContextDropdown.tsx:
```diff
@@ -33,20 +77,24 @@ export const SearchContextDropdown: React.FunctionComponent<SearchContextDropdow
-                    <SearchContextMenu {...props} closeMenu={toggleOpen} />
+                    <SearchContextMenu
+                        {...props}
+                        selectSearchContextSpec={selectSearchContextSpec}
```
suggestion: add unit tests for this functionality

novoselrok commented on client/web/src/integration/search.test.ts:
```diff
@@ -365,4 +365,68 @@ describe('Search', () => {
+            }),
+        }
+
+        test('Streaming search with single repo result', async () => {
```
Git merges are hard 😞

novoselrok commented:
@limitedmage I took your suggestions and added the integration/unit tests. I also noticed that I didn't handle the case if URL context contains an unavailable context spec. So I added the functionality to append it to the query and resubmit the search. Everything is contained here if you want to check it out: https://github.com/sourcegraph/sourcegraph/pull/18209/commits/5ece818313a43fe6767ea218d8febdbd7ff609ff
